keps/sig-node/2371-cri-pod-container-stats/README.md

```diff
@@ -0,0 +1,416 @@
+    // ListPodSandboxStats returns stats of all running pods.
+    rpc ListPodSandboxStats(ListPodSandboxStatsRequest) returns (ListPodSandboxStatsResponse) {}
+    ...
+}
+...
+
+message PodSandboxStatsRequest{
+    // ID of the pod for which to retrieve stats.
+    string pod_id = 1;
+}
+
+message PodSandboxStatsResponse {
+    // Stats of the pod.
+    PodSandboxStats stats = 1;
+}
+
+message ListPodSandboxStatsRequest{
+    // Filter for the list request.
+    PodSandboxStatsFilter filter = 1;
+}
+
+message ListPodSandboxStatsResponse {
+    // Stats of the pod.
+    repeated PodSandboxStats stats = 1;
+}
+
+// PodSandboxStats provides the resource usage statistics for a pod.
+message PodSandboxStats {
+    // Information of the pod.
+    // Corresponds to PodRef in SummaryAPI
+    PodSandboxAttributes attributes = 1;
+    // CPU usage gathered from the pod.
+    // Corresponds to Stats SummaryAPI CPUStats field
+    CpuUsage cpu = 2;
+    // Memory usage gathered from the pod.
+    // Corresponds to Stats SummaryAPI MemoryStats field
+    MemoryUsage memory = 3;
+    // TODO: do we want a start time field?
+    // The time at which data collection for the pod-scoped (e.g. network) stats was (re)started.
+    // int64 timestamp = 1;
+    // Stats of containers in the measured pod.
+    repeated ContainerStats containers
+    // Stats pertaining to CPU resources consumed by pod cgroup (which includes all containers' resource usage and pod overhead).
+    NetworkStats network = 4;
+    // Note the specific omission of VolumeStats and EphemeralStorage
+    // Each of these fields will be calculated Kubelet-level
+    // ProcessStats pertaining to processes.
+    ProcessStats process = 5;
+}
+
+// NetworkStats contains data about network resources.
+message NetworkStats {
+    // The time at which these stats were updated.
+    int64 timestamp = 1;
+
+    // Stats for the default interface, if found
+    InterfaceStats default_interface = 2;
+
+    repeated InterfaceStats interfaces = 3;
+}
+
+// InterfaceStats contains resource value data about interface.
+type InterfaceStats struct {
+    // The name of the interface
+    string name = 1;
+    // Cumulative count of bytes received.
+    Uint64Value rx_bytes = 2;
+    // Cumulative count of receive errors encountered.
+    Uint64Value rx_errors = 2;
+    // Cumulative count of bytes transmitted.
+    Uint64Value tx_bytes = 2;
+    // Cumulative count of transmit errors encountered.
+    Uint64Value tx_errors = 2;
+}
+
+// ProcessStats are stats pertaining to processes.
+message ProcessStats {
+    // Number of processes
+    Uint64Value process_count = 1;
+}
+```
+
+#### Kubelet
+
+Once all required CRI changes are completed, Kubelet can update it's CRI stats provider to stop fetching metrics from cAdvisor and instead obtain the metrics from the CRI for container and pods.
+
+* TODO: How to handle migration to newly created "full" CRI stats provider?
```

Comment:
What about Option 1 but with a feature gate?  I.e if the feature gate is enabled, don't use cAdvisor.  If it is disabled, keep the current behavior.

Reply:
Added as the new option 2